### PR TITLE
add support for nvidia open source driver packages

### PR DESCRIPTION
**Pull request by shivakunv**
@cdesiniotis PTAL.

### CODE REVIEW

File: ubuntu22.04/precompiled/Dockerfile [hunk 1]
```diff
@@ -10,6 +10,7 @@ ENV DRIVER_VERSION=$DRIVER_VERSION
 
 ARG KERNEL_VERSION=5.15.0-60-generic
 ENV KERNEL_VERSION=$KERNEL_VERSION
+ENV USE_OPEN_KERNEL_MODULES=$USE_OPEN_KERNEL_MODULES
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/Dockerfile:13]: This is an env we will set at runtime, not build time, so I believe we can remove this.
```suggestion

```

File: ubuntu22.04/precompiled/Dockerfile [hunk 2]
```diff
@@ -66,6 +67,10 @@ RUN apt-get update && apt-get install -y --no-install-recommends nvidia-driver-$
 RUN apt-get update && apt-get install --download-only --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
     linux-signatures-nvidia-${KERNEL_VERSION} \
     linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}
+# add support for nvidia open source driver packages during runtime    
+RUN apt-get update && apt-get install --download-only --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION} \
+    linux-signatures-nvidia-${KERNEL_VERSION} \
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/Dockerfile:72]: We already install the signatures package in the previous RUN command, so we can remove.
```suggestion
```

**Review by cdesiniotis — COMMENTED** [review]
Left a few minor comments. Otherwise this looks good.

@shivakunv have you been able to verify these changes on an Ubuntu22.04 system? I will test your changes when I find time.

**Comment by shivakunv** [comment]
> Left a few minor comments. Otherwise this looks good.
> 
> @shivakunv have you been able to verify these changes on an Ubuntu22.04 system? I will test your changes when I find time.

I am not able to verify the change.

File: ubuntu22.04/precompiled/nvidia-driver [hunk 3]
```diff
@@ -4,6 +4,7 @@
 set -eu
 
 KERNEL_VERSION=$(uname -r)
+USE_OPEN_KERNEL_MODULES="${USE_OPEN_KERNEL_MODULES:-false}"
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/nvidia-driver:7]: Let's change the variable name to `OPEN_KERNEL_MODULES_ENABLED` to align with our non-precompiled images: https://github.com/NVIDIA/gpu-driver-container/blob/main/ubuntu22.04/nvidia-driver#L19

By using the same variable name, your changes will be compatible with older GPU Operator versions that use this variable name.

File: ubuntu22.04/precompiled/nvidia-driver [hunk 4]
```diff
@@ -210,10 +211,19 @@ _unload_driver() {
 
 # Link and install the kernel modules from a precompiled packages
 _install_driver() {
-    echo "Installing NVIDIA driver kernel modules..."
-    apt-get install --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
-    linux-signatures-nvidia-${KERNEL_VERSION} \
-	linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}
+    if [ "$USE_OPEN_KERNEL_MODULES" = true ]; then
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/nvidia-driver:214]: Let's change the variable name to OPEN_KERNEL_MODULES_ENABLED to align with our non-precompiled images: https://github.com/NVIDIA/gpu-driver-container/blob/main/ubuntu22.04/nvidia-driver#L19

**Review by cdesiniotis — APPROVED** [review]

File: ubuntu22.04/precompiled/Dockerfile [hunk 5]
```diff
@@ -65,7 +65,10 @@ RUN apt-get update && apt-get install -y --no-install-recommends nvidia-driver-$
 # avoid cleaning the cache after this to retain these packages during runtime.
 RUN apt-get update && apt-get install --download-only --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
     linux-signatures-nvidia-${KERNEL_VERSION} \
-    linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}
+    linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
+    # add support for nvidia open source driver packages during runtime
+    linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION} \
+    linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/Dockerfile:71]: As discussed over slack, we should only need the `linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}` package.
```suggestion
    # add support for nvidia open source driver packages during runtime
    linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}
```

File: ubuntu22.04/precompiled/nvidia-driver [hunk 6]
```diff
@@ -210,10 +211,26 @@ _unload_driver() {
...
+        # Installing linux-signatures-nvidia-${KERNEL_VERSION} and linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}
+        # This auto-installs linux-modules-${KERNEL_VERSION}-generic and conflicts with linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}
+        # Using dpkg -i --force-overwrite to resolve the conflict
+        # https://forums.developer.nvidia.com/t/ubuntu-and-nvidia-provided-packages-conflict-breaking-installation/259150
+        apt-get install --download-only --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION} && \
+            dpkg -i --force-overwrite $(find /var/cache/apt/archives/ -name "linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}_*.deb")
+        apt-get install --download-only --no-install-recommends -y linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION} && \
+            dpkg -i --force-overwrite $(find /var/cache/apt/archives/ -name "linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}_*.deb")
+        # Ensure modules are correctly loaded
+        depmod -a ${KERNEL_VERSION}
```
Comment (cdesiniotis) [inline ubuntu22.04/precompiled/nvidia-driver:226]:
```suggestion
        apt-get install --no-install-recommends -y \
            linux-signatures-nvidia-${KERNEL_VERSION} \
            linux-modules-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION}
```

Comment (shivakunv) [reply ubuntu22.04/precompiled/nvidia-driver:226]: linux-signatures-nvidia-${KERNEL_VERSION}   also have conflict with open  linux-modules-nvidia* driver

```
dpkg-deb -c linux-signatures-nvidia-5.15.0-112-generic_5.15.0-112.122+1_amd64.deb
drwxr-xr-x root/root         0 2024-06-14 02:41 ./
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-390/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-450srv/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-470srv/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-535srv/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/nvidia-drm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/nvidia-modeset.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/nvidia-peermem.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/nvidia-uvm.ko.sig
-rw-r--r-- root/root       790 2024-06-14 02:41 ./lib/modules/5.15.0-112-generic/kernel/nvidia-550srv/bits/nvidia.ko.sig
drwxr-xr-x root/root         0 2024-06-14 02:41 ./usr/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./usr/share/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./usr/share/doc/
drwxr-xr-x root/root         0 2024-06-14 02:41 ./usr/share/doc/linux-signatures-nvidia-5.15.0-112-generic/
-rw-r--r-- root/root       443 2024-06-14 02:41 ./usr/share/doc/linux-signatures-nvidia-5.15.0-112-generic/changelog.Debian.gz
-rw-r--r-- root/root     11251 2024-06-14 02:41 ./usr/share/doc/linux-signatures-nvidia-5.15.0-112-generic/copyright
```

Comment (shivakunv) [reply ubuntu22.04/precompiled/nvidia-driver]: verified and It is working fine

File: ubuntu22.04/precompiled/Dockerfile [hunk 7]
```diff
@@ -65,7 +65,10 @@ RUN apt-get update && apt-get install -y --no-install-recommends nvidia-driver-$
 # avoid cleaning the cache after this to retain these packages during runtime.
 RUN apt-get update && apt-get install --download-only --no-install-recommends -y linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
     linux-signatures-nvidia-${KERNEL_VERSION} \
-    linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}
+    linux-modules-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION} \
+    # add support for nvidia open source driver packages during runtime
+    linux-objects-nvidia-${DRIVER_BRANCH}-server-open-${KERNEL_VERSION} \
```
Comment (tariq1890) [inline ubuntu22.04/precompiled/Dockerfile:70]:
```suggestion
```

Comment (cdesiniotis) [reply ubuntu22.04/precompiled/Dockerfile]: As discussed on slack, we need the `linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}` package.

Comment (shivakunv) [reply ubuntu22.04/precompiled/Dockerfile]: updated the code. 
removed linux-objects-nvidia-${DRIVER_BRANCH}-**server-open**-${KERNEL_VERSION} \

Comment (cdesiniotis) [reply ubuntu22.04/precompiled/Dockerfile]: We need to add to add the `linux-objects-nvidia-${DRIVER_BRANCH}-server-${KERNEL_VERSION}` package back.

Comment (cdesiniotis) [reply ubuntu22.04/precompiled/Dockerfile]: Ah nevermind, I see that is still being installed.

**Review by tariq1890 — APPROVED** [review]

**Review by cdesiniotis — APPROVED** [review]

**Pull request merged**